server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,437 @@
+            this::setForcedAwarenessAttributes
+        );
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    private void setForcedAwarenessAttributes(Settings forceSettings) {
+        Map<String, List<String>> forcedAwarenessAttributes = new HashMap<>();
+        Map<String, Settings> forceGroups = forceSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : forceGroups.entrySet()) {
+            List<String> aValues = entry.getValue().getAsList("values");
+            if (aValues.size() > 0) {
+                forcedAwarenessAttributes.put(entry.getKey(), aValues);
+            }
+        }
+        this.forcedAwarenessAttributes = forcedAwarenessAttributes;
+    }
+
+    /**
+     * Starts the new decommission request and registers the metadata with status as {@link DecommissionStatus#INIT}
+     * or the last known status if not {@link DecommissionStatus#FAILED}
+     * Once the status is updated, it tries to exclude to-be-decommissioned cluster manager nodes from Voting Configuration
+     *
+     * @param decommissionAttribute register decommission attribute in the metadata request
+     * @param listener register decommission listener
+     */
+    public synchronized void startDecommissionAction(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        // validates if correct awareness attributes and forced awareness attribute set to the cluster before starting action
+        validateAwarenessAttribute(decommissionAttribute, awarenessAttributes, forcedAwarenessAttributes);
+
+        // register the metadata with status as DECOMMISSION_INIT as first step
+        clusterService.submitStateUpdateTask("decommission [" + decommissionAttribute + "]", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                Metadata metadata = currentState.metadata();
+                Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                // check if the same attribute is requested for decommission and currently not FAILED or SUCCESS, then return the current
+                // state as is
+                if (decommissionAttributeMetadata != null
+                    && decommissionAttributeMetadata.decommissionAttribute().equals(decommissionAttribute)
+                    && !decommissionAttributeMetadata.status().equals(DecommissionStatus.FAILED)
+                    && !decommissionAttributeMetadata.status().equals(DecommissionStatus.SUCCESSFUL)) {
+                    logger.info("re-request received for decommissioning [{}], will not update state", decommissionAttribute);
```

Comment:
should we allow concurrent request for same attribute ? It can lead to some unexpected edge cases. Disallowing it makes it easier for debugging any issues.  We can throw exception that `decommission already in progress` .

Reply:
we cannot do this.. the service is built in a way that it retries itself until cluster manager times out or there’s a hard failure like new attribute requested now or quorum checks etc. The response to the user is sent back once leader is decommissioned
if we don’t allow in progress requests and return the response, then entire self retry mechanism is lost

Reply:
@imRishN : when the status is updated to Successful, is there anything else pending. We can call listener.onResponse() here itself for Successful.

Now coming to failed, lets say previous attempt failed, should you reset the status here to INIT, so that i can be retried? Though it will not maintain any state of the previous attempt here now.

Reply:
> @imRishN : when the status is updated to Successful, is there anything else pending. We can call listener.onResponse() here itself for Successful.

Makes sense. If the same attribute is request for decommission and status is already Success, we can simply acknowledge the request with listener.onResponse
In case, Successful status is for a different attribute, we can reject the request with listener.onFailure

Reply:
> Now coming to failed, lets say previous attempt failed, should you reset the status here to INIT, so that i can be retried? Though it will not maintain any state of the previous attempt here now.

This is what implemented currently. If the previous status is FAILED, and another decommission request is received, we again start with INIT

Reply:
I ideal way we should handle concurrent in progress requests is to fail other requests if decommission metadata is `IN_PROGRESS` while ensuring we are giving a sufficient timeout to the currently running `IN_PROGRESS` and marking it `FAILED` if unable to finish with a finite time. That should allow another request to retried